catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergS3SignParams.java

```diff
@@ -0,0 +1,193 @@
+import org.projectnessie.catalog.service.api.SnapshotResponse;
+import org.projectnessie.error.NessieContentNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.IcebergView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+abstract class IcebergS3SignParams {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IcebergS3SignParams.class);
+
+  abstract IcebergS3SignRequest request();
+
+  abstract ParsedReference ref();
+
+  abstract ContentKey key();
+
+  abstract String baseLocation();
+
+  abstract CatalogService catalogService();
+
+  abstract RequestSigner signer();
+
+  abstract S3BucketOptions s3options();
+
+  @Value.Lazy
+  String requestedS3Uri() {
+    return S3Utils.asS3Location(request().uri());
+  }
+
+  @Value.Lazy
+  boolean write() {
+    return request().method().equalsIgnoreCase("PUT")
+        || request().method().equalsIgnoreCase("POST")
+        || request().method().equalsIgnoreCase("DELETE")
+        || request().method().equalsIgnoreCase("PATCH");
+  }
+
+  Uni<IcebergS3SignResponse> verifyAndSign() {
+    return fetchSnapshot()
+        .call(this::checkForbiddenLocations)
+        .onItem()
+        .transformToMulti(this::collectAllowedLocations)
+        .filter(requestedS3Uri()::startsWith)
+        .toUni()
+        .onItem()
+        .ifNull()
+        .failWith(this::unauthorized)
+        .replaceWith(request().uri())
+        .map(this::sign);
+  }
+
+  private Uni<SnapshotResponse> fetchSnapshot() {
+    try {
+      CompletionStage<SnapshotResponse> stage =
+          catalogService()
+              .retrieveSnapshot(
+                  SnapshotReqParams.forSnapshotHttpReq(ref(), "iceberg", null), key(), null);
+      // consider an import failure as a non-existing content:
+      // signing will be authorized for the future location only.
+      return Uni.createFrom().completionStage(stage).onFailure().recoverWithNull();
+    } catch (NessieContentNotFoundException ignored) {
+      return Uni.createFrom().nullItem();
+    } catch (NessieNotFoundException e) {
+      return Uni.createFrom().failure(e);
+    }
+  }
+
+  private Uni<?> checkForbiddenLocations(SnapshotResponse snapshotResponse) {
+    if (snapshotResponse != null && write()) {
+      Content content = snapshotResponse.content();
+      String metadataLocation = null;
+      if (content instanceof IcebergTable) {
+        metadataLocation = ((IcebergTable) content).getMetadataLocation();
+      } else if (content instanceof IcebergView) {
+        metadataLocation = ((IcebergView) content).getMetadataLocation();
+      }
+      if (metadataLocation != null
+          && requestedS3Uri().equals(S3Utils.normalizeS3Scheme(metadataLocation))) {
+        return Uni.createFrom().failure(unauthorized());
+      }
```

Comment:
Mind adding a TODO to disallow all table and view metadata objects?
I know, if you add the code now, some tests fill fail (the ones that write to temporary locations).